src/SparkTestBase.sol

```diff
@@ -40,22 +45,134 @@ interface IExecutable {
+
+        baseBridge = OptimismBridgeTesting.createNativeBridge(domains[ChainIdUtils.Ethereum()], domains[ChainIdUtils.Base()]);
+
+        allDomains.push(domains[ChainIdUtils.Ethereum()]);
+        allDomains.push(domains[ChainIdUtils.Base()]);
+        allDomains.push(domains[ChainIdUtils.Gnosis()]);
+    }
+
+    function spellIdentifier(ChainId chainId) private view returns(string memory){
+        string memory slug            = string(abi.encodePacked("Spark", chainId.toDomainString(), "_", id));
+        string memory spellIdentifier = string(abi.encodePacked(slug, ".sol:", slug));
+        return spellIdentifier;
+    }
+
+    function deployPayload(ChainId chainId) internal onChain(chainId) returns(address) {
+        return deployCode(spellIdentifier(chainId));
+    }
+
+    function deployPayloads() internal {
+        for (uint256 i = 0; i < allDomains.length; i++) {
+            ChainId id = ChainIdUtils.fromDomain(allDomains[i]);
+            string memory spellIdentifier = spellIdentifier(id);
+            try vm.getCode(spellIdentifier) {
+                payloads[id] = deployPayload(id);
+            } catch {
+                console.log("skipping spell deployment for network: ", id.toDomainString());
+            }
+        }
+    }
+
+    /// @dev takes care to revert the selected fork to what was chosen before
+    function executeAllPayloadsAndBridges() internal {
+        // only execute mainnet payload
+        executePayload(ChainIdUtils.Ethereum());
+        // then use bridges to execute other chains' payloads
+        relayMessageOverBridges();
+    }
+
+    /// @dev bridge contracts themselves are stored on mainnet
+    function relayMessageOverBridges() private onChain(ChainIdUtils.Ethereum()) {
+        // Base
+        OptimismBridgeTesting.relayMessagesToDestination(baseBridge, false);
+        OptimismBridgeTesting.relayMessagesToSource(baseBridge, false);
+        // Gnosis: TODO
+    }
+
+    function executeL2PayloadsFromBridges() private onChain(ChainIdUtils.Ethereum()) {
+        // Base
+        domains[ChainIdUtils.Base()].selectFork();
+        IExecutor(Base.SPARK_EXECUTOR).execute(IExecutor(Base.SPARK_EXECUTOR).actionsSetCount() - 1);
+        // Gnosis: TODO
+    }
+
+    function executePayload(ChainId chain) internal onChain(chain){
+        address payloadAddress = payloads[chain];
+        IExecutor executor = executors[chain];
+        require(Address.isContract(payloadAddress), "PAYLOAD IS NOT A CONTRACT");
+        bool success;
+        if(chain == ChainIdUtils.Ethereum()) {
+            vm.prank(Ethereum.PAUSE_PROXY);
+            (success,) = address(executor).call(abi.encodeWithSignature(
+                'exec(address,bytes)',
+                payloadAddress,
+                abi.encodeWithSignature('execute()')
+            ));
+            require(success, "FAILED TO EXECUTE PAYLOAD");
+        } else {
+            // TODO: research whether it makes sense to prank the same address
```

Comment:
This TODO no longer makes sense if we get rid of the arbitrary-payload execution and only explicitly run the mainnet one

Reply:
Yes I would drop this section.

Reply:
implemented in ca1a3f1

Reply:
lgtm